File: swift/Tests/WalletProviderTests/WalletProviderAttestationOfflineTest.swift
import XCTest
import WalletAPI
import SdJwt
import WalletTestKit
@testable import WalletProvider

/// Server-independent test of the reference adapter's HTTP dance — a fake `HttpTransport` canned to look
/// like the wallet-provider backend, so the sequence and payloads are verified without a running server.
final class WalletProviderAttestationOfflineTest: XCTestCase {

    private actor FakeBackend: HttpTransport {
        private(set) var requests: [HttpRequest] = []
        private var nonces = 0

        func execute(_ request: HttpRequest) async throws -> HttpResponse {
            requests.append(request)
            let json: String
            if request.url.hasSuffix("/nonce") { nonces += 1; json = "{\"nonce\":\"n\(nonces)\"}" }
            else if request.url.hasSuffix("/wallet-instances") { json = "{\"instanceId\":\"inst-1\"}" }
            else if request.url.hasSuffix("/wallet-attestation") { json = "{\"wallet_attestation\":\"header.payload.sig\"}" }
            else if request.url.hasSuffix("/key-attestation") { json = "{\"key_attestation\":\"ka.ka.ka\"}" }
            else { return HttpResponse(status: 404, headers: [], body: []) }
            return HttpResponse(status: 200, headers: [("content-type", "application/json")], body: [UInt8](json.utf8))
        }

        func paths(_ base: String) -> [String] { requests.map { String($0.url.dropFirst(base.count)) } }
        func bodyOf(_ path: String) throws -> JsonValue {
            try JsonValue.parse(String(decoding: requests.first(where: { $0.url.hasSuffix(path) })!.body!, as: UTF8.self))
        }
        func registrationCount() -> Int { requests.filter { $0.url.hasSuffix("/wallet-instances") }.count }
    }

    private func str(_ json: JsonValue?) -> String? { if case let .str(v)? = json { return v }; return nil }

    func testRegistersOnceThenFetchesWuaWithSignedPop() async throws {
        let area = SoftwareSecureArea()
        let key = try await area.createKey(spec: KeySpec(secureArea: area.id))
        let backend = FakeBackend()
        let provider = WalletProviderAttestation(
            baseUrl: "http://wp.test", http: backend, secureArea: area,
            integrity: DevIntegrityTokenProvider(), clientId: "wallet-dev", clock: { 1000 })

        let wua = try await provider.walletAttestation(keyInfo: key)
        XCTAssertEqual(wua, "header.payload.sig")

        // nonce → register → nonce → wallet-attestation
        let paths = await backend.paths("http://wp.test")
        XCTAssertEqual(paths, ["/nonce", "/wallet-instances", "/nonce", "/wallet-attestation"])

        let register = try await backend.bodyOf("/wallet-instances")
        XCTAssertEqual(str(register["integrityToken"]), "dev-integrity:n1")
        XCTAssertEqual(str(register["instanceKey"]?["x"]), Base64Url.encode(key.publicKey.x))

        let attest = try await backend.bodyOf("/wallet-attestation")
        XCTAssertEqual(str(attest["instanceId"]), "inst-1")
        XCTAssertEqual(str(attest["clientId"]), "wallet-dev")
        XCTAssertEqual(str(attest["pop"])?.split(separator: ".").count, 3)

        // A second WUA reuses the cached registration.
        _ = try await provider.walletAttestation(keyInfo: key)
        let registrations = await backend.registrationCount()
        XCTAssertEqual(registrations, 1)
    }
}

File: android/attestation/src/test/kotlin/com/hopae/eudi/wallet/android/attestation/WalletProviderAttestationOfflineTest.kt
package com.hopae.eudi.wallet.android.attestation

import com.hopae.eudi.wallet.sdjwt.Base64Url
import com.hopae.eudi.wallet.sdjwt.JsonValue
import com.hopae.eudi.wallet.spi.HttpRequest
import com.hopae.eudi.wallet.spi.HttpResponse
import com.hopae.eudi.wallet.spi.HttpTransport
import com.hopae.eudi.wallet.spi.KeySpec
import com.hopae.eudi.wallet.testkit.SoftwareSecureArea
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Server-independent unit test of the reference adapter's HTTP dance — a fake [HttpTransport] canned to look
 * like the wallet-provider backend. Runs in normal `./gradlew test` (no server, no `EUDI_WP_LIVE`); the live
 * integration test is the opt-in extra.
 */
class WalletProviderAttestationOfflineTest {
    private class FakeBackend : HttpTransport {
        val requests = mutableListOf<HttpRequest>()
        private var nonces = 0
        override suspend fun execute(request: HttpRequest): HttpResponse {
            requests += request
            val json = when {
                request.url.endsWith("/nonce") -> """{"nonce":"n${++nonces}"}"""
                request.url.endsWith("/wallet-instances") -> """{"instanceId":"inst-1"}"""
                request.url.endsWith("/wallet-attestation") -> """{"wallet_attestation":"header.payload.sig"}"""
                request.url.endsWith("/key-attestation") -> """{"key_attestation":"ka.ka.ka"}"""
                else -> return HttpResponse(404, emptyList(), ByteArray(0))
            }
            return HttpResponse(200, listOf("content-type" to "application/json"), json.encodeToByteArray())
        }

        fun bodyOf(path: String): JsonValue.Obj =
            JsonValue.parse(requests.first { it.url.endsWith(path) }.body!!.decodeToString()) as JsonValue.Obj

        fun paths(base: String) = requests.map { it.url.removePrefix(base) }
    }

    @Test
    fun `registers once, then fetches a WUA with a signed instance-key PoP`() = runBlocking {
        val area = SoftwareSecureArea()
        val key = area.createKey(KeySpec(secureArea = area.id))
        val backend = FakeBackend()
        val provider = WalletProviderAttestation("http://wp.test", backend, area, DevIntegrityTokenProvider(), "wallet-dev")

        val wua = provider.walletAttestation(key)
        assertEquals("header.payload.sig", wua)

        // The Type-4-like sequence: nonce → register → nonce → wallet-attestation.
        assertEquals(listOf("/nonce", "/wallet-instances", "/nonce", "/wallet-attestation"), backend.paths("http://wp.test"))

        // Registration carried the dev integrity token (bound to the first nonce) and the instance-key JWK.
        val register = backend.bodyOf("/wallet-instances")
        assertEquals("dev-integrity:n1", (register["integrityToken"] as JsonValue.Str).value)
        assertEquals(Base64Url.encode(key.publicKey.x), ((register["instanceKey"] as JsonValue.Obj)["x"] as JsonValue.Str).value)

        // wallet-attestation carried the instanceId, the (non-instance-unique) clientId, and a 3-part PoP JWT.
        val attest = backend.bodyOf("/wallet-attestation")
        assertEquals("inst-1", (attest["instanceId"] as JsonValue.Str).value)
        assertEquals("wallet-dev", (attest["clientId"] as JsonValue.Str).value)
        assertEquals(3, (attest["pop"] as JsonValue.Str).value.split(".").size)

        // A second WUA reuses the cached registration — no second POST /wallet-instances.
        provider.walletAttestation(key)
        assertEquals(1, backend.requests.count { it.url.endsWith("/wallet-instances") })
    }
}
